ice.c

```diff
@@ -4484,6 +4484,21 @@ static gboolean janus_ice_outgoing_traffic_handle(janus_ice_handle *handle, janu
 					/* ... but only if this isn't a retransmission (for those we already set it before) */
 					header->ssrc = htonl(video ? stream->video_ssrc : stream->audio_ssrc);
 				}
+				/* Set the abs-send-time value, if needed */
+				if(video && stream->abs_send_time_ext_id > 0) {
+					struct timeval tv;
+					gettimeofday(&tv, NULL);
+					uint64_t s = tv.tv_sec + 2208988800u;
+					uint32_t u = tv.tv_usec;
+					uint32_t f = (u << 12) + (u << 8) - ((u * 3650) >> 6);
+					uint64_t ntp_ts = (s << 32) + f;
+					uint64_t abs24 = (ntp_ts >> 14) & 0x00ffffff;
+					uint32_t abs_ts = abs24;
+					if(janus_rtp_header_extension_set_abs_send_time(pkt->data, pkt->length,
+							stream->abs_send_time_ext_id, abs_ts) < 0) {
+						JANUS_LOG(LOG_ERR, "[%"SCNu64"] Error setting abs-send-time value...\n", handle->handle_id);
```

Comment:
imo the documentation on webrtc is misleading here. It says something about the NTP timestamp but this is not what libwebrtc does:
https://source.chromium.org/chromium/chromium/src/+/main:third_party/webrtc/modules/rtp_rtcp/source/rtp_sender_egress.cc;l=226;drc=6a0a55907bf29c7d6bfedb0f8aa4405ed3e2dcf5
It takes the current monotonic clock and then does some shifting with it:
https://source.chromium.org/chromium/chromium/src/+/main:third_party/webrtc/modules/rtp_rtcp/source/rtp_header_extensions.h;l=42;drc=1f5325b50d27ed221d4a26530908c4b68b834507

i'd strongly recommend the same. Mistakes in this are subtle but may lead to overshooting the available bandwidth

Reply:
Should be fixed now :+1: